rust/otap-dataflow/crates/admin/src/telemetry.rs

```diff
@@ -0,0 +1,867 @@
+            } else {
+                collect_metrics_snapshot(&state.metrics_registry)
+            };
+
+            let response = MetricsWithMetadata {
+                timestamp,
+                metric_sets,
+            };
+
+            Ok(Json(response).into_response())
+        }
+        OutputFormat::JsonCompact => {
+            let metric_sets = if q.reset {
+                collect_compact_snapshot_and_reset(&state.metrics_registry)
+            } else {
+                collect_compact_snapshot(&state.metrics_registry)
+            };
+
+            let response = AllMetrics {
+                timestamp,
+                metric_sets,
+            };
+            Ok(Json(response).into_response())
+        }
+        OutputFormat::LineProtocol => {
+            let body = if q.reset {
+                format_line_protocol(&state.metrics_registry, true, Some(now.timestamp_millis()))
+            } else {
+                format_line_protocol(&state.metrics_registry, false, Some(now.timestamp_millis()))
+            };
+            let mut resp = body.into_response();
+            let _ = resp.headers_mut().insert(
+                header::CONTENT_TYPE,
+                header::HeaderValue::from_static("text/plain; charset=utf-8"),
+            );
+            Ok(resp)
+        }
+        OutputFormat::Prometheus => {
+            let body = if q.reset {
+                format_prometheus_text(&state.metrics_registry, true, Some(now.timestamp_millis()))
+            } else {
+                format_prometheus_text(&state.metrics_registry, false, Some(now.timestamp_millis()))
+            };
+            let mut resp = body.into_response();
+            // Prometheus text exposition format 0.0.4
+            let _ = resp.headers_mut().insert(
+                header::CONTENT_TYPE,
+                header::HeaderValue::from_static("text/plain; version=0.0.4; charset=utf-8"),
+            );
+            Ok(resp)
+        }
+    }
+}
+
+/// Handler for the /telemetry/metrics/aggregate endpoint.
+/// Aggregates metrics by metric set name and optionally by a list of attributes.
+///
+/// Query parameters:
+/// - `reset` (bool, default true): whether to reset metrics after reading.
+/// - `attrs` (string, optional): comma-separated list of attribute names to group by.
+/// - `format` (string, default "json"): output format, one of "json", "json_compact", "line_protocol", "prometheus".
+pub async fn get_metrics_aggregate(
+    State(state): State<AppState>,
+    Query(q): Query<AggregateQuery>,
+) -> Result<Response, StatusCode> {
+    let now = chrono::Utc::now();
+    let timestamp = now.to_rfc3339();
+
+    // Parse attribute list (comma-separated), trim whitespace, drop empties
+    let group_attrs: Vec<String> = q
+        .attrs
+        .as_deref()
+        .unwrap_or("")
+        .split(',')
+        .map(|s| s.trim())
+        .filter(|s| !s.is_empty())
+        .map(|s| s.to_string())
```

Comment:
We should be able to avoid the `to_string()` allocations here. The type of `group_attrs` can be `Vec<&str>`.

```suggestion
        
```

Reply:
Good catch, I just submitted a fix.